docs/adr/0003-consensus-runtime-token-transfer.md

```diff
@@ -0,0 +1,460 @@
+  contains the balance of tokens which are managed exclusively by the runtime
+  and do not belong to any specific regular account in the consensus layer.
+
+  It is not possible to transfer directly into a runtime account and doing so
+  may result in funds to be locked without a way to reclaim them.
+
+  The only way to perform any operations on runtime accounts is through the use
+  of messages emitted by the runtime during each round. These messages are
+  subject to discrepancy detection and instruct the consensus layer what to do.
+
+Combined, the two mechanisms enable account holders to set an allowance in the
+benefit of runtimes so that the runtimes can withdraw up to the allowed amount
+from the account holder's address.
+
+### Addresses
+
+This proposal introduces the following new address context for the runtime
+accounts:
+
+```
+oasis-core/address: runtime
+```
+
+Initial version for the address context is `0`. To derive the address, the
+standard address derivation scheme is used, with the runtime's 32-byte
+identifier used as the `data` part.
+
+### State
+
+This proposal introduces/updates the following consensus state in the staking
+module:
+
+#### General Accounts
+
+The general account data structure is modified to include an additional field
+storing the allowances as follows:
+
+```golang
+type GeneralAccount struct {
+    // ... existing fields omitted ...
+
+    Approvals map[Address]*quantity.Quantity `json:"approvals,omitempty"`
+}
+```
+
+### Transaction Methods
+
+This proposal adds the following new transaction methods in the staking module:
+
+#### Approve
+
+Approve enables an account holder to set an allowance for a beneficiary.
+
+**Method name:**
+
+```
+staking.Approve
+```
+
+**Body:**
+
+```golang
+type Approve struct {
+    Beneficiary  Address           `json:"beneficiary"`
+    Negative     bool              `json:"negative,omitempty"`
+    AmountChange quantity.Quantity `json:"amount_change"`
+}
+```
+
+**Fields:**
+
+- `beneficiary` specifies the beneficiary account address.
+- `amount_change` specifies the absolute value of the amount of base units to
+  change the allowance for.
+- `negative` specifies whether the `amount_change` should be subtracted instead
+  of added.
+
+The transaction signer implicitly specifies the general account. Upon executing
+the approval the following actions are performed:
+
+- If the `disable_allowances` staking consensus parameters is set to `true`,
+  the method fails with `ErrForbidden`.
+
+- The account indicated by the signer is loaded.
+
+- If the approval would create a new allowance and the maximum number of
+  allowances for an account has been reached, the method fails with
+  `ErrInvalidArgument`.
+
+- The set of approvals is updated so that the allowance is updated as specified
+  by `amount_change`/`negative`. In case the change would cause the allowance to
+  be equal to zero or negative, the allowance is removed.
```

Comment:
this will be the only place where we succeed if an amount asked for was too much. are there any specific use cases this leniency is meant to support?

Reply:
So the thing is that you specify how much the allowance changes. So it could happen that you look at the allowance and see that it is X and you want to set it to zero, so you submit a tx T1 `{amount_change: X, negative: true}`. But before that tx is executed, tx T2 is executed which withdraws Y < X. In this case T1 would fail instead of setting the allowance to zero which may not be desired.